apps/internal/base/internal/local/server.go

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+// Package local contains a local HTTP server used with interactive authentication.
+package local
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+)
+
+var okPage = []byte(`
+<!DOCTYPE html>
+<html>
+<head>
+    <meta charset="utf-8" />
+    <title>Authentication Complete</title>
+</head>
+<body>
+    <p>Authentication complete. You can return to the application. Feel free to close this browser tab.</p>
+</body>
+</html>
+`)
+
+const failPage = `
+<!DOCTYPE html>
+<html>
+<head>
+    <meta charset="utf-8" />
+    <title>Authentication Failed</title>
+</head>
+<body>
+	<p>Authentication failed. You can return to the application. Feel free to close this browser tab.</p>
+	<p>Error details: error %s error_description: %s</p>
+</body>
+</html>
+`
+
+// Result is the result from the redirect.
+type Result struct {
+	// Code is the code sent by the authority server.
+	Code string
+	// Err is set if there was an error.
+	Err error
+}
+
+// Server is an HTTP server.
+type Server struct {
+	// Addr is the address the server is listening on.
+	Addr     string
+	resultCh chan Result
+	s        *http.Server
+	reqState string
+}
+
+// New creates a local HTTP server and starts it.
+func New(reqState string) (*Server, error) {
+	var l net.Listener
+	var err error
+	for i := 0; i < 10; i++ {
+		l, err = net.Listen("tcp", "localhost:0")
+		if err != nil {
+			continue
+		}
+		break
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	serv := &Server{
+		Addr:     "http://" + l.Addr().String(),
+		s:        &http.Server{Addr: "localhost:0"},
```

Comment:
Not sure what is happening here ? Does this mean Addr in the line above will be different that on this line ? We would want them to be the same right ?

Reply:
When you use a client, you need to have the calling method (such as http:// or https://). 

When you do the server, it works differently.  It doesn't need to know the protocol, it gets that from other settings.  It just needs to know what to listen on.

So http.Server.Addr is so the server knows what to listen on (and doesn't need the method).
Server.Addr is to give our client the whole string it needs to dial the server.